packages/hydrogen/src/analytics-manager/AnalyticsProvider.tsx

```diff
@@ -0,0 +1,524 @@
+): void;
+function publish(
+  event: typeof AnalyticsEvent.CART_UPDATED,
+  payload: CartUpdatePayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.PRODUCT_ADD_TO_CART,
+  payload: CartLineUpdatePayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.PRODUCT_REMOVED_FROM_CART,
+  payload: CartLineUpdatePayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.CUSTOM_EVENT,
+  payload: OtherData,
+): void;
+function publish(event: any, payload: any): void {
+  if (!areRegistersReady()) {
+    waitForReadyQueue.set(event, payload);
+    return;
+  }
+
+  publishEvent(event, payload);
+}
+
+function publishEvent(event: any, payload: any): void {
+  (subscribers.get(event) ?? new Map()).forEach((callback, subscriber) => {
+    try {
+      callback(payload);
+    } catch (error) {
+      if (typeof error === 'object' && error instanceof Error) {
+        console.error(
+          'Analytics publish error',
+          error.message,
+          subscriber,
+          error.stack,
+        );
+      } else {
+        console.error('Analytics publish error', error, subscriber);
+      }
+    }
+  });
+}
+
+function register(key: string) {
+  if (!registers.hasOwnProperty(key)) {
+    registers[key] = false;
+  }
+
+  return {
+    ready: () => {
+      registers[key] = true;
+
+      if (areRegistersReady() && waitForReadyQueue.size > 0) {
+        waitForReadyQueue.forEach((queuePayload, queueEvent) => {
+          publishEvent(queueEvent, queuePayload);
+        });
+        waitForReadyQueue.clear();
+      }
+    },
+  };
+}
+
+// This functions attempts to automatically determine if the user can be tracked if the
+// customer privacy API is available. If not, it will default to false.
+function shopifyCanTrack() {
+  if (
+    typeof window !== 'undefined' &&
+    typeof window?.Shopify === 'object' &&
+    typeof window?.Shopify?.customerPrivacy === 'object' &&
+    typeof window?.Shopify?.customerPrivacy?.userCanBeTracked === 'function'
+  ) {
+    return window.Shopify.customerPrivacy.userCanBeTracked();
+  }
+  return false;
+}
+
+function AnalyticsProvider({
+  canTrack: customCanTrack,
+  cart: currentCart,
+  children,
+  consent,
+  customData = {},
+  shop: shopProp = null,
+}: AnalyticsProviderProps): JSX.Element {
+  const listenerSet = useRef(false);
+  const {shop} = useShopAnalytics(shopProp);
+  const [consentLoaded, setConsentLoaded] = useState(
+    customCanTrack ? true : false,
+  );
+  const [carts, setCarts] = useState<Carts>({cart: null, prevCart: null});
+  const [canTrack, setCanTrack] = useState(
+    customCanTrack ? () => customCanTrack : () => shopifyCanTrack,
+  );
+
+  // Force a re-render of the value when
+  useEffect(() => {
+    if (customCanTrack) return;
+    if (listenerSet.current) return;
+    listenerSet.current = true;
+
+    document.addEventListener('visitorConsentCollected', () => {
+      setConsentLoaded(true);
+      setCanTrack(() => shopifyCanTrack);
```

Comment:
I'm probably reading this wrong but... shouldn't this be `setCanTrack(shopifyCanTrack)` instead? Or do we need a function returning a function at this point? -- I understand this is needed in `useState` but I think it's not needed here?

Reply:
will clean up

Reply:
yea JP is right - React doesn't like it when we took out the `() =>`